### PR TITLE
(Cgt 2754) - Updated showClientsList to fetch clients list from AgentService

### PR DESCRIPTION
**Pull request by euptonqa**

### CODE REVIEW

File: app/controllers/AgentController.scala [hunk 1]
```diff
@@ -20,26 +20,42 @@ import javax.inject.{Inject, Singleton}
...
-    //TODO remove this dummy code - for test purposes only
-    val clients: Seq[String] = Seq("Client Company 1", "Client Company 2", "Client Individual 3")
-    Future.successful(Ok(views.html.clientList(appConfig, clients)))
+  val showClientList: Action[AnyContent] = authorisedActions.authorisedAgentAction {
+    implicit user =>
+      implicit request =>
+        def handleGGResponse(response: GovernmentGatewayResponse): Result = {
+          response match {
+            case SuccessGovernmentGatewayResponse(clients) => {
+              if (clients.size > 0)
```
Comment (miranda-hawkes) [inline app/controllers/AgentController.scala]: Could you use `.nonEmpty` here so scalastyle doesn't moan please?

File: app/controllers/AgentController.scala [hunk 2]
```diff
@@ -20,26 +20,42 @@ import javax.inject.{Inject, Singleton}
...
-  val showClientList: Action[AnyContent] = Action.async { implicit request =>
-    //TODO remove this dummy code - for test purposes only
-    val clients: Seq[String] = Seq("Client Company 1", "Client Company 2", "Client Individual 3")
-    Future.successful(Ok(views.html.clientList(appConfig, clients)))
+  val showClientList: Action[AnyContent] = authorisedActions.authorisedAgentAction {
+    implicit user =>
+      implicit request =>
+        def handleGGResponse(response: GovernmentGatewayResponse): Result = {
+          response match {
+            case SuccessGovernmentGatewayResponse(clients) => {
```
Comment (miranda-hawkes) [inline app/controllers/AgentController.scala]: Please remove redundant braces

File: app/controllers/AgentController.scala [hunk 3]
```diff
@@ -20,26 +20,42 @@ import javax.inject.{Inject, Singleton}
...
+              if (clients.size > 0)
+                Ok(views.html.clientList(appConfig, clients))
+              else Ok(views.html.confirmPermission(appConfig))
+            }
+            case FailedGovernmentGatewayResponse => InternalServerError
+          }
+        }
+
+        for {
```
Comment (miranda-hawkes) [inline app/controllers/AgentController.scala]: Doing `agentService.getExistingClients(user.authContext).map` might be more appropriate here as there is only one expression in the for {} yield

File: app/views/clientList.scala.html [hunk 4]
```diff
@@ -1,4 +1,4 @@
-@(appConfig: config.AppConfig, clients: Seq[String])(implicit request: Request[_], messages: Messages)
+@(appConfig: config.AppConfig, clients: Seq[models.Client])(implicit request: Request[_], messages: Messages)
```
Comment (miranda-hawkes) [inline app/views/clientList.scala.html]: Could you import models at the top to remove models. please?

File: test/controllers/AgentControllerSpec.scala [hunk 5]
```diff
@@ -28,14 +35,52 @@ import play.api.mvc.{Action, AnyContent, Results}
...
+import play.api.libs.json.Json
+import uk.gov.hmrc.play.http.{HttpGet, HttpPost, HttpPut, HttpResponse}
 
-class AgentControllerSpec extends ControllerSpecHelper {
+import scala.concurrent.Future
+
+class AgentControllerSpec extends ControllerSpecHelper with BeforeAndAfter {
+
+  lazy val injector: Injector = app.injector
+  lazy val appConfig: ApplicationConfig = injector.instanceOf[ApplicationConfig]
```
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: You don't need to set this up - you can use `config` in ControllerSpecHelper

File: test/controllers/AgentControllerSpec.scala [hunk 6]
```diff
@@ -28,14 +35,52 @@ import play.api.mvc.{Action, AnyContent, Results}
...
+  lazy val auditLogger: Logging = injector.instanceOf[Logging]
+  lazy val mockWSHttp: WSHttp = mock[WSHttp]
+
+  before {
+    reset(mockWSHttp)
+  }
+
+
+
+  def mockGovernmentGatewayConnector(governmentGatewayResponse: GovernmentGatewayResponse): GovernmentGatewayConnector ={
```
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: Having a def for this and calling it each time seems as much code as doing `when(mockConnector.getExistingClients(...))` at the start of each test (like with mockWSHttp) as there is no extra logic in the method - it makes it easier to understand the data setup too

File: test/controllers/AgentControllerSpec.scala [hunk 7]
```diff
@@ -55,14 +100,25 @@ class AgentControllerSpec extends ControllerSpecHelper {
         .thenReturn(Action.async(Results.Redirect(testOnlyUnauthorisedLoginUri)))
     }
 
-    new AgentController(config, mockActions, messagesApi)
+    val authModel = mock[AuthorisationDataModel]
```
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: I think the mocking of enrolments response should be done in `mockAuthorisationService()` as it is all part of the Auth service

File: test/controllers/AgentControllerSpec.scala [hunk 8]
```diff
@@ -55,14 +100,25 @@ class AgentControllerSpec extends ControllerSpecHelper {
...
+    new AgentController(mockActions, agentService, appConfig, messagesApi)
   }
 
 
   "Calling .declaration" when {
 
     "provided with a valid authorised user" should {
-      lazy val controller = setupController()
+      val ggConnector = mockGovernmentGatewayConnector(SuccessGovernmentGatewayResponse(Seq(Client("John Smith", List[IdentifierForDisplay]
+        (IdentifierForDisplay("Individual", "johnsmith"))),
```
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: Could you make a val for the sequence of users at the top please? To make the tests easier to read

File: test/controllers/AgentControllerSpec.scala [hunk 9]
```diff
@@ -77,12 +133,80 @@ class AgentControllerSpec extends ControllerSpecHelper {
...
+
+    "provided with an unauthorised user" should {
+      val ggConnector = mockGovernmentGatewayConnector(SuccessGovernmentGatewayResponse(Seq(Client("John Smith", List[IdentifierForDisplay]
+        (IdentifierForDisplay("Individual", "johnsmith"))),
+        Client("Company 123", List[IdentifierForDisplay]
+          (IdentifierForDisplay("Organisation", "company123"))))))
+      val agentService = new AgentService(ggConnector)
+      lazy val controller = setupController(correctAuthentication = false, agentService = agentService)
+      lazy val result = controller.showClientList(FakeRequest())
+      "return a status type of 330" in {
```
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: Status should be 303

File: test/controllers/AgentControllerSpec.scala [hunk 10]
```diff
@@ -28,14 +35,52 @@ import play.api.mvc.{Action, AnyContent, Results}
 import play.api.test.FakeRequest
 import traits.ControllerSpecHelper
 import uk.gov.hmrc.play.frontend.auth.AuthContext
+import org.scalatest.BeforeAndAfter
+import play.api.http.Status._
+import play.api.libs.json.Json
+import uk.gov.hmrc.play.http.{HttpGet, HttpPost, HttpPut, HttpResponse}
```
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: Please remove unused imports

File: test/controllers/AgentControllerSpec.scala [hunk 11]
```diff
@@ -55,14 +100,25 @@ class AgentControllerSpec extends ControllerSpecHelper {
...
   "Calling .declaration" when {
 
     "provided with a valid authorised user" should {
-      lazy val controller = setupController()
+      val ggConnector = mockGovernmentGatewayConnector(SuccessGovernmentGatewayResponse(Seq(Client("John Smith", List[IdentifierForDisplay]
+        (IdentifierForDisplay("Individual", "johnsmith"))),
+        Client("Company 123", List[IdentifierForDisplay]
+          (IdentifierForDisplay("Organisation", "company123"))))))
+      val agentService = new AgentService(ggConnector)
+      lazy val controller = setupController(agentService = agentService)
```
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: I don't think you need to name the agentService parameter here (and in other tests) as it is self-explanatory

File: test/controllers/AgentControllerSpec.scala (hunk 5 above)
Comment (euptonqa) [inline test/controllers/AgentControllerSpec.scala]: I'm injecting ApplicationConfig, not servicesConfig/AppConfig (this caused some confusion for me before)

File: test/controllers/AgentControllerSpec.scala (hunk 11 above)
Comment (euptonqa) [inline test/controllers/AgentControllerSpec.scala]: I think it needs to be explicitly stated what parameter you're assigning to as there are a bunch of default parameters that I don't assign too

File: test/controllers/AgentControllerSpec.scala (hunk 5 above)
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: The AgentController expects an instance of AppConfig to be injected - hence in ControllerSpecHelper it goes `app.injector.instanceOf[AppConfig]` - we're using the config created here in the same way in the subscription controllers

File: test/controllers/AgentControllerSpec.scala [hunk 12]
```diff
@@ -102,21 +102,28 @@ class AgentControllerSpec extends ControllerSpecHelper with BeforeAndAfter {
...
+      .thenReturn(SuccessGovernmentGatewayResponse(clients))}
+
   "Calling .declaration" when {
 
     "provided with a valid authorised user" should {
-      val ggConnector = mockGovernmentGatewayConnector(SuccessGovernmentGatewayResponse(Seq(Client("John Smith", List[IdentifierForDisplay]
-        (IdentifierForDisplay("Individual", "johnsmith"))),
-        Client("Company 123", List[IdentifierForDisplay]
-          (IdentifierForDisplay("Organisation", "company123"))))))
+      ggSetUp
```
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: You don't need this lazy val as you're already mocking the `getExistingClients()` method. Please can you remove `mockGovernmentGatewayConnector()` too and just put the mocking directly in each test i.e.
```
    when(mockConnector.getExistingClients(ArgumentMatchers.any(), ArgumentMatchers.any())(ArgumentMatchers.any()))
      .thenReturn(x)
```
replace x with either the failed response or success with client list

File: test/controllers/AgentControllerSpec.scala [hunk 13]
```diff
@@ -62,14 +62,14 @@ class AgentControllerSpec extends ControllerSpecHelper with BeforeAndAfter {
...
-  def mockAuthorisationService(enrolmentsResponse: Option[Seq[Enrolment]], authResponse: Option[AuthorisationDataModel]): Unit = {
+  def mockAuthorisationService(authResponse: Option[AuthorisationDataModel]): Unit = {
     val mockConnector = mock[AuthorisationConnector]
 
     when(mockConnector.getAuthResponse()(ArgumentMatchers.any()))
       .thenReturn(Future.successful(authResponse))
 
     when(mockConnector.getEnrolmentsResponse(ArgumentMatchers.any())(ArgumentMatchers.any()))
-      .thenReturn(Future.successful(enrolmentsResponse))
+      .thenReturn(Future.successful(Some(Seq(mock[Enrolment]))))
```
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: You don't need to mock anything here - just create a dummy sequence of Enrolments

File: test/controllers/AgentControllerSpec.scala [hunk 14]
```diff
@@ -102,21 +102,28 @@ class AgentControllerSpec extends ControllerSpecHelper with BeforeAndAfter {
 
     val authModel = mock[AuthorisationDataModel]
```
Comment (miranda-hawkes) [inline test/controllers/AgentControllerSpec.scala]: I'm not sure why this needs to be mocked if it's just a case class

**Comment by miranda-hawkes** [comment]
Sorry Emma, one of the test fails. You might need to put the mocking of GG connector and passing this into the AgentService into a def

**Pull request merged**